### PR TITLE
Prevent duplicate GDS emails

### PR DESCRIPTION
**Pull request by pdeziel**
### Scope of changes

Currently GDS allows the same email address to be specified on more than one of a VASP's contacts. This updates the methods on the email client so that they don't send multiple emails to the same address.

Also included is some refactoring to make contact iteration more useable and some basic generated fixtures for the "model" contacts (not the VASP contacts, the ones in their own namespace).

### Type of change

- [x] bug fix
- [ ] new feature
- [ ] documentation
- [ ] other (describe)

### Acceptance criteria

Please review the logic to make sure the contact iteration behavior is the same and the changes to `Register` make sense.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [x]  I have added new test fixtures as needed to support added tests
- [x]   Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [x]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.

### CODE REVIEW

File: fixtures/datagen/fakerize.py [hunk 1]
```diff
@@ -46,6 +46,10 @@
     "Individual",
     "Other",
 ]
+FAKE_CONTACTS = {
+    "adam@example.com": False,
+    "bruce@example.com": True,
+}
```
Comment (pdeziel) [inline fixtures/datagen/fakerize.py:52]: Added contacts fixtures here so we can use them in the GDS test suite.

File: pkg/gds/gds.go [hunk 2]
```diff
@@ -254,31 +262,19 @@ func (s *GDS) Register(ctx context.Context, in *api.RegisterRequest) (out *api.R
...
 			}
-
-			out = &api.RegisterReply{
-				Id:                  vasp.Id,
-				RegisteredDirectory: vasp.RegisteredDirectory,
-				CommonName:          vasp.CommonName,
-				Status:              vasp.VerificationStatus,
-				Message:             "email successfully verified and verification review sent to TRISA admins",
-			}
-			return out, nil
```
Comment (pdeziel) [inline pkg/gds/gds.go:280]: This was a bug I think, if we return here then the certificate request doesn't get created. So I changed this to continue through the contact loop but we should only call `beginReview` once per VASP so we don't spam the TRISA admins.

File: pkg/gds/gds.go [hunk 3]
```diff
@@ -235,6 +229,20 @@ func (s *GDS) Register(ctx context.Context, in *api.RegisterRequest) (out *api.R
...
+		if err = models.SetContactVerification(vaspContact, contact.Token, contact.Verified); err != nil {
+			log.Error().Err(err).Str("contact", kind).Str("vasp", vasp.Id).Msg("could not set contact verification token")
+			return nil, status.Error(codes.Aborted, "could not send contact verification emails")
+		}
+
+		// Prevent sending duplicate verification emails, e.g. if the same email is
+		// specified on multiple contacts in the register request.
+		if _, ok := contacts[contact.Email]; ok {
+			log.Debug().Str("email", contact.Email).Str("kind", kind).Msg("ignoring duplicate email on VASP")
+			continue
```
Comment (pdeziel) [inline pkg/gds/gds.go:243]: Using continue logic here to ensure that we don't send multiple emails.

File: pkg/gds/gds_test.go [hunk 4]
```diff
@@ -177,8 +177,135 @@ func (s *gdsTestSuite) TestRegister() {
...
+			From:      s.svc.GetConf().Email.ServiceEmail,
+			Subject:   emails.VerifyContactRE,
+			Reason:    "verify_contact",
+			Timestamp: sent,
+		},
+	}
+	emails.CheckEmails(s.T(), messages)
+}
+
+func (s *gdsTestSuite) TestRegisterAlreadyVerified() {
```
Comment (pdeziel) [inline pkg/gds/gds_test.go:199]: I think we were missing some coverage for the case where the VASP already has a verified contact so I added it.

File: pkg/models/v1/contacts.go [hunk 5]
```diff
@@ -114,21 +114,56 @@ func contactOrder(contacts *pb.Contacts) []*contactType {
...
 // Returns a new ContactIterator which has Next() and Value() methods that can be used
-// to iterate over contacts in a Contacts object.
-func NewContactIterator(contacts *pb.Contacts, requireEmail bool, requireVerified bool) *ContactIterator {
-	return &ContactIterator{
-		email:    requireEmail,
-		verified: requireVerified,
+// to iterate over contacts in a Contacts object. By default, the iterator will return
+// each contact which is non-zero. SkipNoEmail(), SkipUnverified(), and
+// SkipDuplicates() can be used to filter contacts returned by the iterator.
+func NewContactIterator(contacts *pb.Contacts, opts ...ContactIterOption) *ContactIterator {
```
Comment (pdeziel) [inline pkg/models/v1/contacts.go:155]: Hopefully this is cleaner and less error prone, let me know what you think.

**Review by rebeccabilbro — APPROVED** [review]
I took a pass through and this update makes sense to me @pdeziel -- and thank you for adding the new test coverage, which nicely illustrates how the new deduplication logic works. Are there any additional log messages you'd like to add so that we can monitor how the bug fix is working post-deployment?

**Pull request merged**